[1단계 - 자동차 경주 구현] 우르(김현우) 미션 제출합니다.

java-saeng opened the pull request:
안녕하세요 우아한테크코스 5기 우르입니다.

많이 부족하지만 제가 이번 미션을 진행하면서 생각했던 부분을 적어보겠습니다.

고민한 점
이번 미션을 진행하면서 가장 큰 고민을 했던 부분은 "내가 나중에 이 코드를 바꾸게 되면 어느 부분을 먼저 볼까?"에 대해서 고민을 하며 코드를 작성했습니다.
예를 들어, console 출력의 format을 변경해야한다면 "나는 아마 OutputView 클래스를 먼저 볼 것 같아"라는 생각에 출력 format 책임은 OutputView에서 진행했습니다.

시간 내서 리뷰해주셔서 감사합니다 !

java-saeng committed:
test : (StringTest)

java-saeng committed:
test : (SetTest)

java-saeng committed:
docs : 기능 명세서 추가

java-saeng committed:
feat : (CarInfoValidation) 자동차 이름 입력 예외 추가

feat : (CarInfoValidation) 자동차 이름 입력 예외 추가

feat : (CarInfoValidation) 자동차 이름 입력 예외 추가

java-saeng committed:
feat : (TryCountValidation) 자동차 개수가 음수인지 확인하는 예외 작성

java-saeng committed:
feat : 자동차 정보 입력 받기

java-saeng committed:
feat : Car 도메인 추가

java-saeng committed:
feat : Cars 도메인 추가

java-saeng committed:
feat : RandomNumberGenerator 추가

java-saeng committed:
feat : 자동차 게임 구현 완료

java-saeng committed:
refactor : 구분자 상수 처리

java-saeng committed:
refactor : 거리 default value 상수 처리

java-saeng committed:
refactor : 현재 상태 만드는 메서드 분리

java-saeng committed:
refactor : 구분자 상수 처리

java-saeng committed:
refactor : 매직넘버 상수 처리

java-saeng committed:
fix : 자동차 이름은 5자 이하로 만들어야함

java-saeng committed:
refactor : 매직넘버 상수화

java-saeng committed:
feat : 자동차 중복 이름 예외 추가

java-saeng committed:
refactor : 자동차 현재 출력 상태 책임을 OutputView로 전가

- Cars.getCurrentStatus 테스트 코드 변경

java-saeng committed:
refactor : pr 제출 전 code formatting

jnsorn commented on README.md:
```diff
@@ -5,3 +5,25 @@
+  - , 구분자로 구분
+  - validation 호출
+    - 이름 5자 이하
+    - 차 0대인지 아닌지
```
🤔 기능 명세서가 실제 프로그램과 다른 것 같아요.
빠진 부분은 없는지 확인해볼까요? (ex. 자동차 이름은 중복되지 않아야한다.)

java-saeng replied:
기능 명세서 수정을 생각하지 못했어요,,

말씀대로 기능 명세서를 애플리케이션과 같게 작성하였고, 기능 명세서 구현 여부를 확인하기 위해 체크 박스까지 추가해보았습니다~~

jnsorn replied:
네 좋습니다. 코드와 기능 명세서를 같이 관리한다는게 쉽지는 않겠지만, 프로그램에 대한 명세이니만큼 꼼꼼하게 챙겨보면 좋을 것 같아요. 😃

jnsorn commented on src/main/java/domain/Car.java:
```diff
@@ -0,0 +1,27 @@
+
+    private int distance;
+
+    public Car(final String name) {
+        this.name = name;
+        this.distance = DEFAULT_DISTANCE_VALUE;
+    }
```
해당 생성자를 통해 생성된 자동차는 10자의 이름도 가능하겠네요!
도메인에 도메인 규칙이 적용되어 있지 않네요. 어떻게 고칠 수 있을까요?

java-saeng replied:
말씀대로 도메인에 도메인 규칙이 적용되어 있지 않다고 생각합니다!! 그래서 입력에서 자동차 이름의 길이를 제한하고, 생성자에서도 제한한다면 도메인 규칙이 보다 더 엄격하게 지켜질 것이라 생각해요.

하지만 저는 이 부분에 대해서 제 의견을 말씀드리자면,,  생성자는 객체를 생성하는 아주 무거운 책임을 가지고 있다고 생각해요. 
그래서 생성자에서는 객체 생성을 하는 로직만을 가지고 있는게 좋지 않을까? 라는 생각이 들었습니다.

물론 Car 클래스 안에 validation 메서드를 따로 생성해서, 생성자 내부에서 호출하여 검증할 수 있지만, 이 또한 생성자가 validation 에 관한 책임을 가지고 있고, validation에 대한 명세가 변경된다면 도메인 내부 코드가 변경되어야 할 것 같아요.

결론적으로 Car 클래스 객체를 생성하기 위해서는 무조건 "input 값"이 필요하게 되므로 InputView에서 validation을 처리하게 만들었습니다.

이 부분에 대해서 또링은 어떻게 생각하시는지 궁금합니다~~

java-saeng replied:
만약 제 의견에서 Car 생성자 내부에 validation 로직이 존재하는 것이 생성자 책임이 가중된다고 생각한다면, 
정적 팩토리 메서드를 통해서 해결할 수도 있겠네요!

하지만 Car 클래스에서 validation을 왜 해야할까에 대한 의문은 가지고 있습니다,,ㅠㅠ

jnsorn replied:
우르의 생각을 공유해줘서 고마워요 😎

> 하지만 저는 이 부분에 대해서 제 의견을 말씀드리자면,, 생성자는 객체를 생성하는 아주 무거운 책임을 가지고 있다고 생각해요.
> 그래서 생성자에서는 객체 생성을 하는 로직만을 가지고 있는게 좋지 않을까? 라는 생각이 들었습니다.
> 물론 Car 클래스 안에 validation 메서드를 따로 생성해서, 생성자 내부에서 호출하여 검증할 수 있지만, 이 또한 생성자가 validation 에 관한 책임을 가지고 있고, validation에 대한 명세가 변경된다면 도메인 내부 코드가 변경되어야 할 것 같아요.

자동차 이름에 대한 검증이라면 자동차 이름이라는 객체를 만들어 책임을 할당해도 괜찮지 않을까요? 😃 자동차라고 해서 자동차가 모든 책임을 갖기보다는, 각 책임이 적절한 객체를 찾아보거나 만들어내는 것도 하나의 객체가 너무 많은 일을 하지 않도록 하는 좋은 방법일 것 같아요. 마치 우르가 자동차 경주 게임을 구현하기 위해 자동차 경주 게임이라는 객체만으로 구현한 것이 아닌 Car와 CarEngine등의 여러 객체들을 협력하여 구현한 것 처럼요!

jnsorn commented on src/main/java/domain/Cars.java:
```diff
@@ -0,0 +1,69 @@
+import java.util.List;
+import java.util.Map;
+
+public class Cars {
+
+    private final List<Car> cars;
```
자동차들의 행위를 모아 일급컬렉션을 생성하셨군요 💯

java-saeng replied:
같이 페어했던 제이의 아이디어 였습니다~~!! 감사합니다 :)

jnsorn commented on src/main/java/domain/Cars.java:
```diff
@@ -0,0 +1,69 @@
+
+    private boolean canMove() {
+        return RandomNumberGenerator.generateRandomNumber() >= 4;
+    }
```
자동차의 움직일 수 있는 조건을 판단하고 움직이는건 cars의 역할일까요?
해당 책임이 더 적절한 객체는 없을까요?

java-saeng replied:
말씀대로 cars는 car에 대한 일급 컬렉션으로 자동차들의 행위를 모아서 관리(?) 하는 객체입니다.
그런데 여기서 자동차를 움직이게 하거나, 움직일 수 있도록 하는 조건들은 다른 곳에서 처리해야한다고 생각해요.

그래서 별도의 CarEngine 이라는 클래스를 생성하여 자동차를 움직이게 하는 doRace, canMove, moveCar 메서드를 분리했습니다 !

jnsorn replied:
엇 제 의도가 조금 다르게 전달된 것 같아요!
저는 자동차가 움직일 수 있는 조건을 외부(cars)에서 판단하고 자동차를 움직이기 보다는, 자동차에게 움직이라는 신호를 주면 자동차가 움직일 수 있는 조건을 판단하여 움직이는 형태를 생각했습니다. (= canMove()를 판단하는건 자동차의 책임에 더 어울리지 않을까하는 생각이었어요) 
별도로 만들어주신 CarEngine은 Cars의 역할과 크게 다를게 없어보이는데요, 우르 생각은 어떠세요?

jnsorn commented on src/main/java/domain/Cars.java:
```diff
@@ -0,0 +1,69 @@
+        if (car.getDistance() == max) {
+            result.add(car.getName());
+        }
+    }
```
만약 우승 자동차의 distance까지 같이 출력해야한다면 이 코드는 어떻게 바뀌게 될까요?
view에 의해 도메인 코드에 변경이 일어날 것 같아요. 😮
도메인 코드가 view에 의한 영향을 받지 않으려면 어떻게 해야 좋을지도 고민해봐요!

java-saeng replied:
말씀대로 우승 자동차의 distance 까지 같이 출력되야한다면 view 변경으로 인해 도메인 코드에 변경이 발생하게 될 것 같아요!!

그래서 Cars 에서는 우승 자동차만을 넘겨주고, OutputView에서 우승 자동차를 받아서 명세에 맞게 변경시키도록 수정했습니다 ~~

jnsorn commented on src/main/java/engine/RacingGameEngine.java:
```diff
@@ -0,0 +1,53 @@
+
+    private static final String SPLIT_DELIMITER = ",";
+
+
```
🔎 컨벤션을 확인해주세요.

java-saeng replied:
말씀하시려는 컨벤션이 혹시 불필요한 공백이 맞는지 궁금합니다~~!!

jnsorn replied:
맞습니다 :)

jnsorn commented on src/main/java/engine/RacingGameEngine.java:
```diff
@@ -0,0 +1,53 @@
+        return new Cars(Arrays.stream(carName.split(SPLIT_DELIMITER))
+                              .map(Car::new)
+                              .collect(Collectors.toList()));
+    }
```
🔎 컨벤션을 확인해주세요.

java-saeng replied:
return 문에 로직들을 포함하지 않고, return에는 꼭 필요로하는 정보만 작성하였습니다!!

혹시 말씀하시는 컨벤션이 맞을까요~~?

https://myeonguni.tistory.com/1596 
7.3번 참고했습니다

jnsorn replied:
엇 요건 제가 잘못 확인한 것 같네요 😅

jnsorn commented on src/main/java/validation/TryCountValidation.java:
```diff
@@ -0,0 +1,12 @@
+
+public class TryCountValidation {
+
+    private static final int BOUNDARY_NEGATIVE_VALUE = 0;
```
0번 이동도 가능한가요? ㅎㅎ
프로그래밍 요구사항에 해당 규칙은 없지만, 게임을 제공하는 입장에서 고민해봐도 재미있겠네요.

java-saeng replied:
"멈춘다"라는 명세를 자동차가 0 거리만큼 움직인다고 생각했습니다 !!

jnsorn replied:
엇 이 부분은 시도횟수에 대한 리뷰였어요! ㅎㅎ

jnsorn commented on src/test/java/domain/CarsTest.java:
```diff
@@ -0,0 +1,59 @@
+        assertThat(carCurrentStatus).hasSize(2)
+                                    .contains(entry("pobi", 2), entry("crong", 1));
+    }
+}
```
(참고) [파일 끝에 개행을 추가해야 하는 이유](https://blog.coderifleman.com/2015/04/04/text-files-end-with-a-newline/ )

java-saeng replied:
오,, 신기하군여,, 

인텔리제이에서 제공해주는 옵션이 있어서 바로 적용해보았습니다!!

jnsorn commented on src/test/java/domain/CarsTest.java:
```diff
@@ -0,0 +1,59 @@
+    @DisplayName("getWinner() : 가장 많이 움직인 자동차를 반환해준다.")
+    void test_getWinner() {
+        // given
+        final Car pobi = createCar("pobi");
```
위에서 언급한 케이스가 여기있네요!

> 해당 생성자를 통해 생성된 자동차는 10자의 이름도 가능하겠네요!
도메인에 도메인 규칙이 적용되어 있지 않네요. 어떻게 고칠 수 있을까요?

java-saeng replied:
> 말씀대로 도메인에 도메인 규칙이 적용되어 있지 않다고 생각합니다!! 그래서 입력에서 자동차 이름의 길이를 제한하고, 생성자에서도 제한한다면 도메인 규칙이 보다 더 엄격하게 지켜질 것이라 생각해요.
 하지만 저는 이 부분에 대해서 제 의견을 말씀드리자면,, 생성자는 객체를 생성하는 아주 무거운 책임을 가지고 있다고 생각해요.
그래서 생성자에서는 객체 생성을 하는 로직만을 가지고 있는게 좋지 않을까? 라는 생각이 들었습니다.
 물론 Car 클래스 안에 validation 메서드를 따로 생성해서, 생성자 내부에서 호출하여 검증할 수 있지만, 이 또한 생성자가 validation 에 관한 책임을 가지고 있고, validation에 대한 명세가 변경된다면 도메인 내부 코드가 변경되어야 할 것 같아요.
 결론적으로 Car 클래스 객체를 생성하기 위해서는 무조건 "input 값"이 필요하게 되므로 InputView에서 validation을 처리하게 만들었습니다.
 이 부분에 대해서 또링은 어떻게 생각하시는지 궁금합니다~~

jnsorn commented on src/test/java/study/SetTest.java:
```diff
@@ -0,0 +1,47 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class SetTest {
+    private Set<Integer> numbers;
+
+    @BeforeEach
+    void setUp() {
+        numbers = new HashSet<>();
+        numbers.add(1);
+        numbers.add(1);
+        numbers.add(2);
+        numbers.add(3);
+    }
+
+    @Test
+    void setSizeTest() {
+        // when, then
+        assertEquals(numbers.size(), 3);
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {"1", "2", "3"})
+    void isContainsSet(int input) {
+        // when, then
+        assertTrue(numbers.contains(input));
+    }
+
+    @ParameterizedTest
+    @CsvSource(value = {"1,true", "2,true", "3,true", "4,false", "5,false"}, delimiter = ',')
+    void isContainsSet(int input, boolean check) {
+        // when, then
+        assertEquals(numbers.contains(input), check);
+    }
```
(이번 리뷰 범위는 아니지만) 꼼꼼하게 학습하셨군요 👍

java-saeng replied:
![image](https://user-images.githubusercontent.com/62413589/218261740-d810120a-e252-4b2b-a11f-8717348df05d.png)

1단계 미션 이전에 있어서 같이 묶음으로 하는 줄 알았습니다,, 하하 그래도 리뷰해주셔서 감사합니다 !!

jnsorn commented on src/test/java/validation/CarInfoValidationTest.java:
```diff
@@ -0,0 +1,71 @@
+
+    @Test
+    @DisplayName("validationCar() : 이름 5자 초과일 경우에 IllegalArgumentException 발생")
+    void test_ValidateName_IllegalArgumentException() {
```
⚠️ 테스트가 실패하고 있어요. 확인해주세요!

java-saeng replied:
중간에 에러 메시지를 변경했었는데 테스트를 같이 수정하지 않았나봅니다,,,
다음부터는 제출 전에 모든 테스트를 돌려보고 제출하는 습관을 갖도록 하겠습니다!

jnsorn commented on src/test/java/validation/CarInfoValidationTest.java:
```diff
@@ -0,0 +1,71 @@
+            final String errorMessage = exception.getMessage();
+
+            assertEquals(expectedMessage, errorMessage);
+        }
```
예외에 대한 테스트를 위해 assertj에서 많은 기능을 제공하고 있어요.
학습자료에도 나와있으니 참고해서 더 명확한 테스트를 작성해볼까요?

java-saeng replied:
![image](https://user-images.githubusercontent.com/62413589/218303820-16160248-925b-4c34-9c00-93bdbfbabf7f.png)

assertThrows를 사용하여 에러 테스트를 진행했었는데 사진과 같이 validateLength, validateName, validateDuplicatedCarName 는 private 메서드이고, validation이 지켜지지 않을경우에 모두 같은 에러를 반환하기 때문에 테스트 하기 어렵다고 생각이 들었습니다. 그래서 try-catch를 통해 에러 핸들링할 때 받은 메시지로 검증을 해보자! 라고 생각을 했어요.

하지만 처음에는 또링이 말씀하신 "명확한 테스트"가 assertThrows인건가? 라는 생각을 했고, 찾아보니 hasMessage 메서드를 알게 됐고, 제 테스트코드에 적용을 해보았습니다 !!

적용하고나서,  "왜 굳이 assertThrow를 사용하지? 내가 작성해놓은 테스크 코드도 명확한 테스트를 할 수 있지 않았을까?"에 대해서 생각해보았어요,,

그런데 "자동차의 개수가 1 이상이어야한다" 이 부분에서 상상하지도 못한 에러가 발생했습니다,,

exception이 터지지도 않아서 catch 문에서 assert문이 실행되지 않은 체로 테스트가 완료되서 성공한 것처럼 되더라구요,, 

그래서 왜 또링이 말씀하신 assertJ를 통한 명확한 테스트를 작성하는지 몸으로 느낄 수 있었어요,, 

긴 글 읽어주셔서 감사해요 :)

jnsorn commented on src/test/java/validation/TryCountValidationTest.java:
```diff
@@ -0,0 +1,31 @@
+        // when & then
+        assertThrows(IllegalArgumentException.class,
+                     () -> TryCountValidation.validateTryCount(input));
+    }
+
+    @Test
+    @DisplayName("validateTryCount() 성공 경우")
+    void test_validateTryCount_success() {
+        // given
+        final int input = 3;
```
성공하는 경우와 실패하는 경우를 잘 나눠 테스트해주셨네요!
특정 숫자보다는 경계값을 테스트하면 더 견고한 테스트 코드를 작성할 수 있어요 ☺️

java-saeng replied:
그렇군요,, 결과에 영향 없는 값 ~ 영향 있는 값 그 사이 경계값을 테스트하면 더 견고한 테스트 코드를 작성할 수 있겠네요!!

그러면 혹시 여기서 말씀하시는 경계값이라 함은 0을 말씀하시는걸까요~?

jnsorn replied:
네 맞습니다 ㅎㅎ 지금 코드는 -1은 불가능, 0은 가능이라는 규칙을 가지고 있으니 이 두 값으로 테스트 해보면 실패하는 경우와 성공하는 경우를 명확하게 테스트 할 수 있을 것 같아요.

jnsorn reviewed (changes requested):
안녕하세요. 우르! 첫 미션인데 빠르게 잘 구현해주셨어요 👍 
특히 변경에 유연한 코드를 계속 고민하면서 작성하셨다는게 멋지네요! 
생각해보면 좋을 것 같은 부분에 코멘트 남겼으니 확인해주세요.
궁금하거나 이해가 안되는 부분 있다면 편하게 DM 주셔도 좋습니다!

java-saeng committed:
docs : 기능 명세서 수정

- "자동차 이름 중복 불가" 추가
- 체크박스 추가

java-saeng committed:
feat : Cars.getWinner에서 List<String>이 아닌 List<Car> 반환

- view에 의해 도메인 코드에 변경이 일어날 수 있어서 OutputView에 Car의 이름이 아닌 Car 객체를 반환
- 그에 따라 Test code 변경

java-saeng committed:
refactor : 컨벤션 지키기 !

- 불필요한 공백 삭제
- return 문에는 필요한 정보만을 작성하고, 로직들은 그 위에 따로 작성하도록 함

java-saeng committed:
test : 에러 메시지가 달라 테스트 실패 수정

java-saeng committed:
fix : "" 나 " " 값이 split에 의해 제대로 걸러지지 않기 때문에 자동채 개수가 1개로 인식됨

- 테스트 수정 중 split은 빈 문자열일 때도 길이가 1인 문자열 배열을 반환하는 것을 알고 커밋 메시지와 같은 상황이 발생함
- 따라서 배열 길이 - 1 이 0이면서 index 가 0인 value 값이 "" 또는 " " 일 경우 예외 반환으로 수정

java-saeng committed:
style : pr 제출 전 reformatting

- 파일 끝에 개행 추가

java-saeng committed:
refactor : car를 움직이는 책임 분리

- 기존 car 를 움직인 것은 Cars 를 통해 움직임
- 하지만 일급 컬렉션의 cars보다는 자동차의 움직일 수 있는 조건을 판단하고 움직이게 하는 책임을 하는 객체 필요
- CarEngine을 통해서 자동차의 움직일 수 있는 조건을 판단하고 움직이게 함

java-saeng commented:
안녕하세요 또링~ 먼저 리뷰 감사해요!

코드 리뷰를 받아보고 평소에 사소하게 생각했던 것들에 대해서 다시 생각해 볼 수 있는 좋은 기회였어요
또링이 말씀해주신 부분에 대해서 제 의견을 남겨보았습니다~~

java-saeng committed:
refactor : 일급 컬렉션에서의 List<Car> 불변성을 보장하기 위해 unmodifiableList로 반환

jnsorn commented on src/main/java/domain/Car.java:
```diff
@@ -0,0 +1,27 @@
+package domain;
+
+public class Car {
```
자동차 경주게임에서 아주 중요한 로직을 담당하고 있는데 단위테스트가 없네요 🥲
모든 도메인에 대해 테스트를 추가해주세요!

jnsorn commented on src/main/java/engine/CarEngine.java:
```diff
@@ -0,0 +1,24 @@
+import domain.Cars;
+import utils.RandomNumberGenerator;
+
+public class CarEngine {
```
여기도 테스트가 없어요 🥲

jnsorn replied:
아마 CarEngine 테스트를 작성하려보면 랜덤값때문에 테스트가 어려울텐데요, 어떻게 해결할 수 있을지 고민해볼까요? 🤔

jnsorn commented on src/main/java/validation/CarInfoValidation.java:
```diff
@@ -0,0 +1,50 @@
+    private static final int CAR_COUNT_MIN = 0;
+    private static final String SPLIT_DELIMITER = ",";
+
+    public static void validateCar(String input) {
+        final String[] cars = input.split(SPLIT_DELIMITER);
+        validateLength(cars);
+        validateName(cars);
+        validateDuplicatedCarName(cars);
+    }
```
CarInfoValidation이라는 이름 아래 여러 책임들이 섞여있는 것 같아요. 
DM으로 나누었던 내용대로 리팩터링 해볼까요? 
(여러 책임들이 섞이니 클래스 이름도 Info라는 아주 추상적인 단어가 사용된 것 같네요.)

jnsorn commented on src/main/java/validation/CarInfoValidation.java:
```diff
@@ -0,0 +1,50 @@
+        validateDuplicatedCarName(cars);
+    }
+
+    private static void validateLength(final String[] cars) {
+        if (cars.length - 1 == CAR_COUNT_MIN && isEmpty(cars[0])) {
+            throw new IllegalArgumentException("자동차를 한 대 이상 작성해주세요.");
+        }
+    }
```
`" ,   ,  ,"` 가 입력되면 어떻게 될까요?

jnsorn reviewed (approved):
안녕하세요. 우르! 피드백 반영 잘 해주셨어요. 👏🏻
조금 더 고민해보면 좋을 부분 코멘트 남겨두었으니 같이 고민하면서 2단계 진행해주세요.
내일부터 새로운 미션이 시작해서 시간이 많이 없을텐데, 너무 부담느끼지 마시고 천천히 우르 속도에 맞춰 구현해주세요! 궁금하거나 빠르게 이야기 나눠보고 싶은 내용은 언제든 편히 DM 주시구요 :)

jnsorn commented:
이번 리뷰에서는 크게 3가지를 중심으로 고민&리팩터링 해주시면 좋을 것 같아요. 😃

- 자동차 이름과 위치에 대한 검증은 어떤 객체의 책임일까?
- CarEngine의 역할은 무엇일까?
- 우리가 제어할 수 없는 부분에 대한 테스트는 어떻게 하면 좋을까?
**아, 그리고 모든 도메인에 대해 단위테스트를 작성해주세요. (중요!)**

아마 내일부터 사다리 미션을 시작해서 시간이 부족할테니, 천천히 수정해주세요. 특히 리뷰의 의도를 모르겠거나 궁금한 부분은 편히 DM주시구요!
